apps/discord-bot-internal/src/services/streams.ts

```diff
@@ -0,0 +1,267 @@
+      .getLiveMomentumModStreams()
+      .catch(() => null);
+
+    if (streams === null) return false;
+
+    await Promise.all(
+      messages.map(async (message) => {
+        try {
+          if (message.embeds.length !== 1) return await message.delete();
+
+          const matchingStream = streams.find(
+            (s) => s.user_name === message.embeds[0].author?.name
+          );
+
+          if (!matchingStream) return await message.delete();
+
+          if (this.streamerIdMessageMap.has(matchingStream.user_id)) {
+            logger.warn(
+              `Duplicate cached streamer: ${matchingStream.user_name}, deleting...`
+            );
+            return await message.delete();
+          }
+
+          this.streamerIdMessageMap.set(matchingStream.user_id, message.id);
+          this.messageIdStreamerMap.set(message.id, matchingStream.user_id);
+        } catch (error) {
+          logger.warn(
+            { err: error },
+            'Failed to parse existing message ' + message.id
+          );
+        }
+      })
+    );
+
+    return true;
+  }
+
+  async deleteBannedMessages(messages: Message[]) {
+    await Promise.all(
+      messages
+        .map((msg) => [msg, this.messageIdStreamerMap.get(msg.id)] as const)
+        .filter(
+          ([_, streamerId]) =>
+            streamerId !== undefined &&
+            config.twitch_user_bans.includes(streamerId)
+        )
+        .map(async ([msg, streamerId]) => {
+          this.messageIdStreamerMap.delete(msg.id);
+          this.streamerIdMessageMap.delete(streamerId!);
+          try {
+            await msg.delete();
+          } catch (error) {
+            logger.warn(
+              { err: error },
+              'Failed to delete message of banned stream ' + msg.id
+            );
+          }
+        })
+    );
+  }
+
+  async processEndedStreams(streams: TwitchStream[], messages: Message[]) {
+    const endedStreams = this.streamerIdMessageMap
+      .entries()
+      .toArray()
+      .filter(
+        ([streamerId]) =>
+          !streams.some((stream) => stream.user_id === streamerId)
+      );
+
+    for (const [streamerId, messageId] of endedStreams) {
+      if (this.softBans.has(streamerId)) this.softBans.delete(streamerId);
+
+      const streamMessage = messages.find((msg) => msg.id === messageId);
+      if (streamMessage)
+        streamMessage
+          .delete()
+          .catch((error) =>
+            logger.warn(
+              { err: error },
+              'Failed to delete message of ended stream ' + streamMessage.id
+            )
+          );
+
+      this.streamerIdMessageMap.delete(streamerId);
+      this.messageIdStreamerMap.delete(messageId);
+    }
+  }
+
+  async registerSoftBans(messages: Message[]) {
+    const selfMessages = (messages = messages.filter(
+      (m) => m.author.id === this.client.user.id
+    ));
+    const softBannedMessages = this.messageIdStreamerMap
+      .entries()
+      .toArray()
```

Comment:
I would generally do the .toArray last. It's partly convention, I don't know how much V8 benefits doing this bit last, but def something I'd yell at people for in C#/LINQ so I'm going to yell about it here as well :)